File: src/wiki/publish.py
"""Static site build orchestration."""

from __future__ import annotations

import shutil
from pathlib import Path

from .assets import build_asset_manifest, write_packaged_asset
from .errors import BuildError
from .paths import build_page_manifest, detect_output_collisions, page_output_path
from .render import render_markdown_files
from .schemas import BuildOptions, BuildResult
from .session import Wiki
from .site import build_index_html, build_page_html, build_site


def _path_is_same_or_ancestor(ancestor: Path, descendant: Path) -> bool:
    ancestor = ancestor.resolve()
    descendant = descendant.resolve()
    if ancestor == descendant:
        return True
    try:
        descendant.relative_to(ancestor)
        return True
    except ValueError:
        return False


def _validate_build_output_dir(page_output_dir: Path, config) -> None:
    page_output_dir = page_output_dir.resolve()
    protected: list[tuple[str, Path]] = [
        ("config root", config.config_root.resolve()),
    ]
    for input_dir in config.wiki.inputs:
        protected.append(("wiki input", input_dir.resolve()))
    for asset_dir in config.wiki.assets:
        protected.append(("wiki asset", asset_dir.resolve()))
    layout = config.page_layout
    if layout is not None and layout.is_file():
        protected.append(("page layout", layout.parent.resolve()))

    for label, path in protected:
        if _path_is_same_or_ancestor(page_output_dir, path):
            raise BuildError(
                f"refusing to clean build output path {page_output_dir} because it "
                f"overlaps {label} at {path}. Choose a separate output directory such as _site."
            )


def _build_static_site(wiki: Wiki, options: BuildOptions) -> BuildResult:
    config = wiki.config
    written_paths: list[Path] = []

    if options.render_first:
        graph = wiki.graph(infer=True, reload=options.reload_graph, disk_cache=options.disk_cache)
        success, _errors, _stale, _render_errors = render_markdown_files(config, graph)
        if options.disk_cache and success > 0:
            wiki.graph(infer=True, reload=True, disk_cache=True)

    if not any(path.exists() for path in config.wiki.inputs):
        dirs_str = ", ".join(str(path) for path in config.wiki.inputs)
        return BuildResult(ok=False, error_message=f"none of the input directories exist ({dirs_str})")

    if not options.skip_preflight:
        preflight = wiki.preflight()
        if preflight.errors or not preflight.ok:
            return BuildResult(ok=False, preflight=preflight)

    base_url = config.site.base_url or ""
    url_style = config.site.url_style or "dir"
    site = build_site(config, base_url=base_url, url_style=url_style)
    output_dir = options.output_dir.resolve()

    config_root = config.config_root
    default_layout: Path | None = None
    if config.page_layout is not None and config.page_layout.is_file():
        default_layout = config.page_layout

    page_output_dir = output_dir / base_url.strip("/") if base_url else output_dir
    manifest = (
        build_page_manifest(config, page_output_dir, base_url, url_style)
        + build_asset_manifest(config, page_output_dir, base_url)
    )
    collision_issues = detect_output_collisions(manifest)
    if collision_issues:
        from .schemas import AuditReport, Issue

        preflight = AuditReport(
            ok=False,
            errors=[
                Issue(code="output_collision", message=message, severity="error")
                for message in collision_issues
            ],
        )
        return BuildResult(ok=False, preflight=preflight)

    page_output_dir = page_output_dir.resolve()
    _validate_build_output_dir(page_output_dir, config)

    if page_output_dir.exists():
        shutil.rmtree(page_output_dir)
    page_output_dir.mkdir(parents=True, exist_ok=True)

    has_root_index = any(page.full_slug == "" for page in site.pages)
    if not has_root_index:
        index_path = page_output_dir / "index.html"
        index_path.write_text(
            build_index_html(
                site,
                config_root,
                base_url=base_url,
                url_style=url_style,
                default_layout=default_layout,
            ),
            encoding="utf-8",
        )
        written_paths.append(index_path.relative_to(output_dir))

    for page in site.pages:
        file_path = page_output_path(page_output_dir, page.full_slug, url_style)
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_text(
            build_page_html(
                page,
                config_root,
                base_url=base_url,
                url_style=url_style,
                default_layout=default_layout,
            ),
            encoding="utf-8",
        )
        written_paths.append(file_path.relative_to(output_dir))

    asset_entries = build_asset_manifest(config, page_output_dir, base_url)
    for entry in asset_entries:
        entry.output_path.parent.mkdir(parents=True, exist_ok=True)
        if entry.source is not None:
            shutil.copy2(entry.source, entry.output_path)
        elif entry.output_path.name.endswith(".css"):
            write_packaged_asset(entry.output_path.name, entry.output_path)
        written_paths.append(entry.output_path.relative_to(output_dir))

    return BuildResult(
        ok=True,
        page_count=len(site.pages),
        asset_count=len(asset_entries),
        written_paths=written_paths,
    )

File: src/wiki/site/publish.ts
import { BuildError } from "../errors.ts";
import { dirname, join, resolve } from "@std/path";
import { isFile, pathExists, relativeWithin } from "../fspath.ts";
import type { Wiki } from "../wiki.ts";

import { buildAssetManifest, iterAssetFiles } from "../assets.ts";
import {
  buildPageManifest,
  detectOutputCollisions,
  pageOutputPath,
} from "../paths.ts";
import {
  AuditReport,
  type BuildOptions,
  type BuildResult,
} from "../schemas/reports.ts";
import { buildIndexHtml, buildPageHtml } from "./html.ts";
import { buildSite } from "./build.ts";

function pathIsSameOrAncestor(ancestor: string, descendant: string): boolean {
  const root = resolve(ancestor);
  const candidate = resolve(descendant);
  if (root === candidate) return true;
  try {
    relativeWithin(candidate, root);
    return true;
  } catch {
    return false;
  }
}

function validateOutputDir(pageOutputDir: string, wiki: Wiki): void {
  const protectedPaths: [string, string][] = [
    ["config root", wiki.config.config_root],
    ...wiki.config.wiki.input.map((path) =>
      ["wiki input", path] as [string, string]
    ),
    ...wiki.config.wiki.assets.map((path) =>
      ["wiki asset", path] as [string, string]
    ),
  ];
  const layout = wiki.config.site.layout;
  if (layout !== null && isFile(layout)) {
    protectedPaths.push(["page layout", dirname(layout)]);
  }
  for (const [label, path] of protectedPaths) {
    if (!pathIsSameOrAncestor(pageOutputDir, path)) continue;
    throw new BuildError(
      `refusing to clean build output path ${
        resolve(pageOutputDir)
      } because it overlaps ${label} at ${
        resolve(path)
      }. Choose a separate output directory such as _site.`,
    );
  }
}

function outputSubdirectory(outputDir: string, baseUrl: string): string {
  const relative = baseUrl.replace(/^\/+|\/+$/g, "");
  const parts = relative.split("/").filter((part) => part !== "");
  if (parts.some((part) => part === "." || part === "..")) {
    throw new BuildError(`invalid site.base_url path: ${baseUrl}`);
  }
  return parts.length === 0 ? outputDir : join(outputDir, ...parts);
}

function relativeOutputPath(path: string, outputDir: string): string {
  return relativeWithin(resolve(path), resolve(outputDir));
}

export async function buildStaticSite(
  wiki: Wiki,
  options: BuildOptions,
): Promise<BuildResult> {
  const config = wiki.config;
  if (options.render_first) {
    const renderOptions = {
      ...(options.reload_graph ? { reload: true } : {}),
      ...(options.disk_cache ? { cache: true } : {}),
    };
    await wiki.render(null, renderOptions);
  }

  if (!config.wiki.input.some((path) => pathExists(path))) {
    const directories = config.wiki.input.join(", ");
    return {
      ok: false,
      page_count: 0,
      asset_count: 0,
      written_paths: [],
      error_message: `none of the input directories exist (${directories})`,
    };
  }

  if (!options.skip_preflight) {
    const preflight = await wiki.preflight();
    if (preflight.errors.length > 0 || !preflight.ok) {
      return {
        ok: false,
        page_count: 0,
        asset_count: 0,
        written_paths: [],
        preflight,
      };
    }
  }

  const baseUrl = options.base_url ?? config.site.base_url;
  const urlStyle = options.url_style ?? config.site.url_style;
  const site = buildSite(config, baseUrl, urlStyle);
  const outputDir = resolve(options.output_dir);
  const pageOutputDir = resolve(outputSubdirectory(outputDir, baseUrl));
  const manifest = [
    ...buildPageManifest(config, pageOutputDir, baseUrl, urlStyle),
    ...buildAssetManifest(config, pageOutputDir, baseUrl),
  ];
  const collisions = detectOutputCollisions(manifest);
  if (collisions.length > 0) {
    return {
      ok: false,
      page_count: 0,
      asset_count: 0,
      written_paths: [],
      preflight: new AuditReport({
        ok: false,
        errors: collisions.map((message) => ({
          code: "output_collision",
          message,
          severity: "error" as const,
        })),
      }),
    };
  }

  validateOutputDir(pageOutputDir, wiki);
  if (pathExists(pageOutputDir)) {
    Deno.removeSync(pageOutputDir, { recursive: true });
  }
  Deno.mkdirSync(pageOutputDir, { recursive: true });

  const defaultLayout =
    config.site.layout !== null && isFile(config.site.layout)
      ? config.site.layout
      : null;
  const writtenPaths: string[] = [];
  if (!site.pages.some((page) => page.file_slug === "")) {
    const indexPath = join(pageOutputDir, "index.html");
    Deno.writeTextFileSync(
      indexPath,
      buildIndexHtml(site, baseUrl, urlStyle, defaultLayout),
    );
    writtenPaths.push(relativeOutputPath(indexPath, outputDir));
  }

  for (const page of site.pages) {
    const outputPath = pageOutputPath(pageOutputDir, page.file_slug, urlStyle);
    Deno.mkdirSync(dirname(outputPath), { recursive: true });
    Deno.writeTextFileSync(
      outputPath,
      buildPageHtml(page, baseUrl, defaultLayout),
    );
    writtenPaths.push(relativeOutputPath(outputPath, outputDir));
  }

  const assetEntries = buildAssetManifest(config, pageOutputDir, baseUrl);
  for (const entry of assetEntries) {
    Deno.mkdirSync(dirname(entry.output_path), { recursive: true });
    if (entry.source !== null) {
      Deno.copyFileSync(entry.source, entry.output_path);
    }
    writtenPaths.push(relativeOutputPath(entry.output_path, outputDir));
  }

  return {
    ok: true,
    page_count: site.pages.length,
    asset_count: iterAssetFiles(config).length,
    written_paths: writtenPaths,
  };
}
